Implement Elytra

pandaaaBE opened the pull request:
## Introduction
Implement **player-focused** Elytra.

(Waiting until someone adds this: https://github.com/pmmp/PocketMine-MP/issues/6124)

_There is a plugin, why should we accept your pull request?
Well, since IvanCrafter is currently working on a Fireworks PR it would be only fitting to finally add the Elytra as well - it's tedious to search out all the vanilla features on poggit._

## Todo
- [x] Add to damage event (kinetic energy damage)

## Tests
[Video](https://streamable.com/iefpxd)

pandaaaBE committed:
add elytra class

pandaaaBE committed:
add to itemtypeids

pandaaaBE committed:
add to stringtoitemparser

pandaaaBE committed:
add to vanillaitems

pandaaaBE committed:
add to ItemSerializerDeserializerRegistrar

pandaaaBE committed:
cs

ShockedPlot7560 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+			$this->applyDamageTime--;
+			if($this->applyDamageTime <= 0){
+				$this->applyDamage(1);
+				$this->applyDamageTime = 61;
```
Where does this hard value of 61 come from?

pandaaaBE replied:
A pair of elytra has 432 durability, allowing 7 minutes and 12 seconds of gliding time - in onTickWorn this is something like a "counter", if u glide 60 secounds then it will apply 1 damage

ShockedPlot7560 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+				$this->applyDamageTime = 61;
+			}
+
+			$location = $entity->getLocation();
```
This variable is useless

pandaaaBE committed:
remove hard value

pandaaaBE committed:
add setFallDistance and changed APPLY_DAMAGE_INTERVAL to 21

setFallDistance: in combination with resetFallDistance you won´t get fall damage, idk why

JavierLeon9966 reviewed (changes requested):
Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.

pandaaaBE commented:
> Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.

the rest is okay?

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+	private const MINIMUM_PITCH = -59;
+	private const MAXIMUM_PITCH = 38;
+
+	private const APPLY_DAMAGE_INTERVAL = 21;
```
Shouldn't this be `20`?

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+			// Check if the player's pitch is within the allowed range between -59 and 38.
+			if ($pitch >= self::MINIMUM_PITCH && $pitch <= self::MAXIMUM_PITCH) {
+				$entity->resetFallDistance();
+				$entity->setFallDistance(0.0);
```
This is redundant. `resetFallDistance()` does the same thing.

pandaaaBE replied:
Tested it with resetFallDistance and I always get damage when gliding to the ground, in combo with setFallDistance I don‘t get fall damage..

What is the reason?

JavierLeon9966 replied:
The reason is probably unrelated to this. I don't see how adding redundancy could fix that.

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+			if ($entity->isOnGround()) {
+				$entity->setGliding(false);
+				$entity->resetFallDistance();
+				$entity->setFallDistance(0.0);
```
Same here.

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+
+			$pitch = $entity->getLocation()->pitch;
+			// Check if the player's pitch is within the allowed range between -59 and 38.
+			if ($pitch >= self::MINIMUM_PITCH && $pitch <= self::MAXIMUM_PITCH) {
```
This should be calculated instead of checking the pitch. You can probably find the calculation in the Java source code ot it's probably if the Y velocity is greater than the XZ velocity.

pandaaaBE replied:
> $velocity = $entity->getMotion();
> $yVelocity = $velocity->y;
> $xzVelocity = sqrt($velocity->x ** 2 + $velocity->z ** 2);
> 
> if ($yVelocity > $xzVelocity) {
> }

Like this? 🤔

JavierLeon9966 replied:
Based on tests fall damage is reset if the vertical velocity is less than 0.5 and those angles are what make it less than that. You can't rely on motion for the velocity because players always have 0 motion server-side.

JavierLeon9966 commented:
> > Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.
> 
> the rest is okay?

A few more things.

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,71 @@
+	private int $applyDamageTime = self::APPLY_DAMAGE_INTERVAL;
+
+	public function onTickWorn(Living $entity) : bool {
+		if ($entity instanceof Human && $entity->isGliding()) {
```
This can extend to `Living` entities.

dktapps commented:
As with all previous PRs on the same subject (there have been many), this just relies on client-side movement, which means that elytra are useless for server-side mobs. It's not possible to, for example, make a gliding NPC.

pandaaaBE committed:
fix fall damage

pandaaaBE committed:
Merge branch 'minor-next' into elytra

pandaaaBE committed:
cs?

pandaaaBE committed:
...

pandaaaBE commented:
> > > Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.
> > 
> > 
> > the rest is okay?
> 
> A few more things.

Is there something like a sample I can use to find out how I can "register" something in EntityDamageEvent?

pandaaaBE commented:
> Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.

![image](https://github.com/pmmp/PocketMine-MP/assets/135014092/3708e5b5-9b7a-4be0-b143-3ef460d5a133)

pandaaaBE committed:
add kinetic energy damage

pandaaaBE committed:
add to playerdeathevent

pandaaaBE commented:
Okay.. maybe kinetic energy is to strong currently.. 

https://streamable.com/pqx40j

JavierLeon9966 commented:
> > > > Another thing is that elytras can give you kinetic energy damage if you go too fast against a wall or roof.
> > > 
> > > 
> > > the rest is okay?
> > 
> > 
> > A few more things.
> 
> Is there something like a sample I can use to find out how I can "register" something in EntityDamageEvent?

Just add a `CAUSE_FLY_INTO_WALL` constant in `EntityDamageEvent`.

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,85 @@
+			$direction = $entity->getDirectionVector();
+			$yVel = $direction->y;
+
+			if($yVel < 0.5){
```
That isn't how you get the velocity. It's supposed to be `currentY - prevY` but there needs to be a way to get the previous position outside of the `Entity` class

pandaaaBE replied:
> It's supposed to be currentY - prevY

Yes, I know but there is no method for it - at least nothing found... Is there anything outside the entity class or you didn´t know?

I could log the previous Position but idk if this is a great idea

pandaaaBE replied:
private $previousPosition = null;

    public function onTickWorn(Living $entity): bool {
        $currentPosition = $entity->getLocation();

        if ($this->previousPosition === null) {
            // Store the current position as the previous position in the first tick
            $this->previousPosition = $currentPosition;
        }

        $deltaY = $currentPosition->getY() - $this->previousPosition->getY();
        $deltaXY = $currentPosition->distance($this->previousPosition);

        // Calculate vertical and horizontal velocities
        $yVelocity = $deltaY;
        $xyVelocity = $deltaXY;

        if ($yVelocity > $xyVelocity) {
        }

        // Update the previous position for the next tick
        $this->previousPosition = $currentPosition;

        return false;
    }

JavierLeon9966 replied:
Just make an issue about it and wait until that method gets added or make a PR that adds it.

pandaaaBE replied:
> ue about it and wait until that method gets added or make a PR that adds it.

would it be bad if i first add myself what i sent above and later add the new method?

JavierLeon9966 commented on src/item/Elytra.php:
```diff
@@ -0,0 +1,85 @@
+			}
+
+			//Calculate kinetic energy.
+			$xzVector = new Vector2($direction->x, $direction->z);
```
Same goes here.

pandaaaBE replied:
can i add the function like in the example above my self or should i wait for a other build-in method?

pandaaaBE committed:
little changes

just the last dev update, but still not tested - would be nice if you review

pandaaaBE reviewed (commented):
Needs to be tested, would this be right with the velocity?

@JavierLeon9966

pandaaaBE committed:
add knowntranslationfactory thing

JavierLeon9966 commented:
> Needs to be tested, would this be right with the velocity?
> 
> @JavierLeon9966

No, I don't think you're supposed to do that with items. Just make an issue about needing a way to get an entity's previous position or position Delta and Dylan will handle it.

pandaaaBE commented:
> > Needs to be tested, would this be right with the velocity?
> > @JavierLeon9966
> 
> No, I don't think you're supposed to do that with items. Just make an issue about needing a way to get an entity's previous position or position Delta and Dylan will handle it.

@JavierLeon9966 - now that Dylan is gone, can we introduce this by getting the previous position in the Elytra class, or do you have an idea how I could introduce this in the code

ShockedPlot7560 commented:
> do you have an idea how I could introduce this in the code

This would be a question to be asked in the issue no ? 
Of course Dylan will be gone, but that doesn't mean we have to start issuing several features in one PR.
